source/src/microapp/cs_MicroappStorage.cpp

```diff
@@ -47,331 +53,335 @@ NRF_FSTORAGE_DEF(nrf_fstorage_t nrf_microapp_storage) =
-}
+	if (isErased(appIndex)) {
+		return ERR_SUCCESS;
+	}
 
-uint16_t MicroappStorage::erasePages() {
-	uint32_t err_code;
-	err_code = nrf_fstorage_erase(&nrf_microapp_storage, nrf_microapp_storage.start_addr, g_FLASH_MICROAPP_PAGES, NULL);
-	return err_code;
+	uint32_t flashAddress = nrf_microapp_storage.start_addr + appIndex * MICROAPP_MAX_SIZE;
+	LOGMicroappInfo("erase addr=0x%08X size=%u", flashAddress, MICROAPP_MAX_SIZE / 0x1000);
+	uint32_t nrfCode = nrf_fstorage_erase(&nrf_microapp_storage, flashAddress, MICROAPP_MAX_SIZE / 0x1000, NULL);
+	if (nrfCode != NRF_SUCCESS) {
+		LOGe("Failed to start erase: %u", nrfCode);
+		return ERR_UNSPECIFIED;
+	}
+	return ERR_WAIT_FOR_SUCCESS;
 }
 
-uint16_t MicroappStorage::checkAppSize(uint16_t size) {
-	uint32_t start = nrf_microapp_storage.start_addr;
-	if ((start + size) > nrf_microapp_storage.end_addr) {
-		LOGw("Microapp binary too large. Application can not be written");
+cs_ret_code_t MicroappStorage::writeChunk(uint8_t appIndex, uint16_t offset, const uint8_t* data, uint16_t size) {
+	uint32_t flashAddress = nrf_microapp_storage.start_addr + appIndex * MICROAPP_MAX_SIZE + offset;
+	LOGMicroappInfo("Write chunk at 0x%08X of size %u", flashAddress, size);
+
+	if (offset + size > MICROAPP_MAX_SIZE) {
 		return ERR_NO_SPACE;
 	}
-	return ERR_SUCCESS;
+
+	if (size % 4 != 0) {
+		LOGw("Chunk size must be multiple of 4B, size=%u", size);
+		return ERR_WRONG_PAYLOAD_LENGTH;
+	}
+
+	if (!isErased(flashAddress, size)) {
+		LOGw("Chunk is not erased");
+		return ERR_WRITE_DISABLED;
+	}
+
+	// Write chunk in even smaller chunks of MICROAPP_STORAGE_BUF_SIZE.
+	// TODO: This could be avoided if `data` would be word aligned,
+	// which we could do by making sure the control command payload is word aligned.
+	_chunkData = data;
+	_chunkSize = size;
+	_chunkWritten = 0;
+	_chunkFlashAddress = flashAddress;
+
+	return writeNextChunkPart();
 }
 
-/**
- * We assume that the checksum of the particular chunk is already performed. It would be a waste to send an event with
- * incorrect checksum. If other sources for microapp code (e.g. UART) are added, the checksum should also be checked
- * early in the process. We also assume that all data buffers are of size MICROAPP_CHUNK_SIZE. The last one should be
- * appended with 0xFF values to make it the right size.
- */
-uint16_t MicroappStorage::writeChunk(uint8_t index, const uint8_t *data, uint16_t size) {
-	uint32_t err_code;
-	uint32_t start = nrf_microapp_storage.start_addr + (MICROAPP_CHUNK_SIZE * index);
-	LOGi("Write chunk: %i at 0x%08X of size %i", index, start, size);
-	LOGi("Start with data [%02X,%02X,%02X]", data[0], data[1], data[2]);
-
-	// check chunk size
-	if ((start + size) > nrf_microapp_storage.end_addr) {
-		LOGw("Microapp binary too large. Chunk not written");
-		return ERR_NO_SPACE;
+cs_ret_code_t MicroappStorage::writeNextChunkPart() {
+	LOGMicroappDebug("writeNextChunkPart _chunkSize=%u _chunkWritten=%u", _chunkSize, _chunkWritten);
```

Comment:
Needs nullptr check. `_buffer` cannot be assumed to be ok for writing to. It is allocated at ::init rather than construction (and may fail if there is a memory shortage).

Reply:
Allocating write buffer at construction.